syncthing-discovery/src/main/kotlin/net/syncthing/java/discovery/protocol/LocalDiscoveryUtil.kt

```diff
@@ -0,0 +1,142 @@
+    private const val INCOMING_BUFFER_SIZE = 1024
+
+    private val logger = LoggerFactory.getLogger(javaClass)
+
+    suspend fun listenForAnnounceMessages(): ReceiveChannel<LocalDiscoveryMessage> = GlobalScope.produce {
+        DatagramSocket(LISTENING_PORT, InetAddress.getByName("0.0.0.0")).use { datagramSocket ->
+            invokeOnClose {
+                datagramSocket.close()
+            }
+
+            withContext(Dispatchers.IO) {
+                val datagramPacket = DatagramPacket(ByteArray(INCOMING_BUFFER_SIZE), INCOMING_BUFFER_SIZE)
+
+                while (!isClosedForSend) {
+                    try {
+                        datagramSocket.receive(datagramPacket)
+                    } catch (ex: SocketException) {
+                        if (datagramSocket.isClosed) {
+                            // if the socket was closed by the invokeOnClose, then ignore it
+
+                            return@withContext
+                        } else {
+                            // otherwise it's more serious and it is rethrown
+
+                            throw ex
+                        }
+                    }
+
+                    try {
+                        val sourceAddress = datagramPacket.address.hostAddress
+                        val byteBuffer = ByteBuffer.wrap(datagramPacket.data, datagramPacket.offset, datagramPacket.length)
+                        val magic = byteBuffer.int
+                        NetworkUtils.assertProtocol(magic == MAGIC) {
+                            "magic mismatch, expected ${MAGIC}, got $magic"
+                        }
+                        val announce = LocalDiscoveryProtos.Announce.parseFrom(ByteString.copyFrom(byteBuffer))
+                        val deviceId = DeviceId.fromHashData(announce.id.toByteArray())
+
+                        val deviceAddresses = (announce.addressesList ?: emptyList()).map { address ->
+                            // When interpreting addresses with an unspecified address, e.g.,
+                            // tcp://0.0.0.0:22000 or tcp://:42424, the source address of the
+                            // discovery announcement is to be used.
+                            DeviceAddress.Builder()
+                                    .setAddress(address.replaceFirst("tcp://(0.0.0.0|):".toRegex(), "tcp://$sourceAddress:"))
+                                    .setDeviceId(deviceId.deviceId)
+                                    .setInstanceId(announce.instanceId)
+                                    .setProducer(DeviceAddress.AddressProducer.LOCAL_DISCOVERY)
+                                    .build()
+                        }
+
+                        val message = LocalDiscoveryMessage(
+                                deviceId = deviceId,
+                                addresses = deviceAddresses
+                        )
+
+                        send(message)
+                    } catch (ex: Exception) {
```

Comment:
again

Reply:
I could reduce it to IOException here (which can be thrown by NetworkUtils.assertProtocol and during parsing the protocol buffers) but in case of any unexpected exception during handling the package, it's better to just ignore that package and log something instead of crashing the whole local discovery